### PR TITLE
Support logging on evaluation end

### LINKED ISSUE #5285: Allow self.log from on_validation_end
## 🚀 Feature

Currently, `self.log` is [not supported](https://github.com/PyTorchLightning/pytorch-lightning/blob/9c8701f2e2c09a984ac0322942c8f742a7cc6bc9/pytorch_lightning/trainer/connectors/logger_connector/callback_hook_validator.py#L194) in `on_validation_end`. I would like it to be.

### Motivation

I want to compute some metric each time the validation loop runs, but I don't want it to be tied to `on_validation_batch_end`: i.e I want to do my own batching. I also want to `self.log` this so I can then use it as a key to monitor for checkpointing. Finally, the documentation claims:

```
Use the log() method to log from anywhere in a LightningModule and Callback except functions with batch_start in their names.
```

I don't see why we should prevent logging in this method, it just adds a surprising friction.

### Pitch

Don't prevent logging in `on_validation_end`.

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #5285

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @carmocca @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy @awaelchli @ninginthecloud @daniellepintz

### CODE REVIEW

File: tests/loops/test_evaluation_loop.py [hunk 1]
```diff
@@ -42,25 +42,6 @@ def test_on_evaluation_epoch_end(eval_epoch_end_mock, tmpdir):
     assert eval_epoch_end_mock.call_count == 4
 
 
-@mock.patch(
-    "pytorch_lightning.trainer.connectors.logger_connector.logger_connector.LoggerConnector.log_eval_end_metrics"
-)
-def test_log_epoch_metrics_before_on_evaluation_end(update_eval_epoch_metrics_mock, tmpdir):
```
Comment (carmocca) [inline tests/loops/test_evaluation_loop.py:48]: This test is not useful anymore because logging results can be computed dynamically. This wasn't the case when it was added: https://github.com/PyTorchLightning/pytorch-lightning/issues/7166

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 2]
```diff
@@ -154,8 +157,8 @@ def on_run_end(self) -> List[_OUT_DICT]:
         # log metrics
         self.trainer.logger_connector.log_eval_end_metrics()
 
-        # hook
-        self._on_evaluation_end()
```
Comment (tchaton) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:158]: Good catch !

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 3]
```diff
@@ -145,6 +145,9 @@ def on_run_end(self) -> List[_OUT_DICT]:
         # hook
         self._on_evaluation_epoch_end()
 
+        # hook
+        self._on_evaluation_end()
```
Comment (rohitgr7) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:149]: just concerned with an edge case where a user logs inside this hook and expect `ModelCheckpoint` or `EarlyStopping` to monitor it, or some other crazy callback using `callback_metrics/logged_metrics`.

where do we draw a line and have a hook where we expect all the logs will be present?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/loops/dataloader/evaluation_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:149]: This is not a problem after the logger redesign because `ModelCheckpoint` saves (for validation) also during `on_validation_end` and it's the last callback to run.

Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:149]: but we run callbacks before lightningmodule:
https://github.com/PyTorchLightning/pytorch-lightning/blob/1a2536361749d13e3f30cec11c7da11747f425ab/pytorch_lightning/loops/dataloader/evaluation_loop.py#L229-L231

Comment (carmocca) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:149]: Mmm great point.

Yeah then this won't work well for the `LightningModule`.

So we have these options:
1. Drop this feature
1. Go forward knowing this
1. Invert the order for this hook in particular (not a great idea as it would be a breaking change).

What do you think is the best? We should probably mention this in the issue.

Comment (rohitgr7) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py:149]: personally 1. since we have a workaround of using `on_validation_epoch_end`. Also, the user seems to be happy with it: https://github.com/PyTorchLightning/pytorch-lightning/issues/5285#issuecomment-1017022560

2. IMO, will leak an issue even though many users maybe not be using it, but we can't tell for sure.
3. same reason.
